### PR TITLE
Use parallel test execution mode in maven-surefire-plugin. #1122

### PR DESCRIPTION
**Pull request by mkordas**
After this change test execution time on my machine has dropped from 30 to 20 seconds.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java [hunk 1]
```diff
@@ -69,7 +69,7 @@
      * key is "line:column"
      * value is DetailNode tree
      */
-    private static final Map<String, ParseStatus> TREE_CACHE = new HashMap<>();
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java]: `static` cache is not multithreading-friendly...

**Comment by romani** [comment]
> private final Map<String, ParseStatus> treeCache = new HashMap<>();

in this case each instance of JavadocCheck will have its own cache. We need to move that cache out of Abstract class.

**Comment by mkordas** [comment]
The checks can share cache, but right now the keys are stored using just `lineNo:columnNo` and two separate files can interfere...

**Comment by romani** [comment]
As far as i remember we clean cache after we processed a file, problem should not happen.
As I put a comment above, i am ok to move cache out to separate class to let all Checks use it, it still has to to cleaned at the end of file processing.

**Comment by mkordas** [comment]
That's right, cache gets cleaned for every file. It works if only one file is processed at once. These surefire options make tests executing in multiple threads and multiple files being processed at once causing flickering tests.

**Comment by romani** [comment]
We will not danage perfomance of Checkstyle production execution to let UT run a bit quicker.

**Comment by mkordas** [comment]
What about https://github.com/checkstyle/checkstyle/wiki/Checkstyle-GSoC-2015-Project-Ideas#project-name-multi-thread-mode-for-java-files-processing? It's one step in that direction.

**Comment by romani** [comment]
MT is huge task that definitely could be solved from multiple sides.
Checkstyle is single threaded by design. That particular idea is attempt to stay single threaded on file processing but be multi-threaded on Validation on a file. That idea could be done in limited amount of time and thoroughly tested, and it is first step to final completely MT implementation.

Smth that you propose is very far idea that is not feasible to archived this year (I mean reliable implementation of it). So I would rather avoid any changes to this idea for now, especially in expense of performance. 

User will use us even we are slow, but they will hate us for bugs, noisy false-positives and unfortunate exceptions/crashes.

**Comment by mkordas** [comment]
Ok, I see your point.

On the other hand making field `private static final` of a mutable, not thread-safe class is dodgy practice anyway.

What about making this field `ThreadLocal`? Users won't feel the difference on a single-threaded execution, we'll be able to test in parallel and our code will follow good practices.

**Comment by romani** [comment]
no changes for now please, I do not want to spend a time on that now. When we come to MT we will think about that. Stability is in more priority now.
We are single threaded for now.

**Pull request closed without merging**